calblueprint/ogsc
# Profile updates: schema changes + reusable profile section and field editor components

⚠️ **Do not merge until the follow-on PR (#205) has been approved and scheduled to merge immediately after!** This diff is split out for review purposes only.

* Removed BMI, Team Name, and Player Number profile fields from players, replacing it with Height and Year of Birth attributes
* Adds migration utility for updating enum members in Postgres: `updateEnumMigration`
* Adds standard validators for each profile field value type in lib/validate
* New reusable components for Profile: ProfileFieldCell (replaces ProfileContentCell), ProfileFieldEditor, and ProfileSection.
  * ProfileFieldCell: continues to act as a renderer for profile fields of varying value types. New changes make it aware of the editing state of its surrounding section and auto-swap in a ProfileFieldEditor if in an editing state.
  * ProfileFieldEditor: renders standard UI for each kind of value type.
  * ProfileSection: surrounds a group of ProfileFieldCells and maintains their editing state. Sections are sensitive to lib/access definitions and determine whether or not an "edit" button should appear alongside the section.
* Adds serialization function for profile fields, updates serialization strategy for DistanceMeasured, TimeElapsed value types
* Database seeding script adds fewer profile field values, to make seed content significantly more lightweight

**Breaking changes:**
* Re-seeding the database is required after this PR is merged. Serialization strategies for certain value types have changed and will result in a broken state for fields typed with DistanceMeasured or TimeElapsed if the DB is not re-seeded. _(In the future/when this ships, we will not be afforded such flexibility with the serialization strategy changing. We should make absolutely sure that profile fields are serialized the way we want now!)_
* Any references to PlayerNumber, TeamName, and BMI will no longer work (they have been dropped).

## TODO

Addressed in follow-on PRs (#205):
* Edit states for the ValueHistoryView-driven components
* Reusing ProfileFieldEditor in the creation form pages
* Type errors caused by incompleteness of certain types across API endpoints and creation form page-related components

## How to Test/Use PR feature
(Read Breaking Changes section for running locally. The Heroku deploy is affected to fail for this particular PR due to type errors mentioned above.)

* Validate that profile field value _updating_ is working as expected across all user role types. The editing behavior/rules should correspond to the definitions in lib/access.

## PR Checklist

Does your branch (check if true):

## Related PRs

- #205: Reuses ProfileFieldEditor in creation forms

## Migrations

* 20210216055344-UpdateProfileFields: Makes profile changes requested by OGSC, mentioned above.

CC: @sydbui
